signer/core/signed_data.go

```
@@ -0,0 +1,714 @@
+	}
+
+	message, ok := raw["message"].(string)
+	messageBytes, err := hexutil.Decode(message)
+	if err != nil {
+		return ValidatorData{}, err
+	}
+	if !ok || len(messageBytes) == 0 {
+		return ValidatorData{}, errors.New("message is undefined")
+	}
+
+	return ValidatorData{
+		Address: common.BytesToAddress(addrBytes),
+		Message: messageBytes,
+	}, nil
+}
+
+// Validate make sure the types are sound
+func (typedData *TypedData) Validate() error {
+	if err := typedData.Types.Validate(); err != nil {
+		return err
+	}
+	if err := typedData.Domain.Validate(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Map generates a map version of the typed data
+func (typedData *TypedData) Map() map[string]interface{} {
+	dataMap := map[string]interface{}{
+		"types":       typedData.Types,
+		"domain":      typedData.Domain.Map(),
+		"primaryType": typedData.PrimaryType,
+		"message":     typedData.Message,
+	}
+	return dataMap
+}
+
+// Validate checks if the types object is conformant to the specs
+func (types *EIP712Types) Validate() error {
+	for typeKey, typeArr := range *types {
+		for _, typeObj := range typeArr {
+			typeVal := typeObj["type"]
+			if typeKey == typeVal {
+				return fmt.Errorf("type '%s' cannot reference itself", typeVal)
+			}
+			firstChar := []rune(typeVal)[0]
+			if unicode.IsUpper(firstChar) {
+				if (*types)[typeVal] == nil {
+					return fmt.Errorf("referenced type '%s' is undefined", typeVal)
+				}
+			} else {
+				if !isStandardTypeStr(typeVal) {
+					if (*types)[typeVal] != nil {
+						return fmt.Errorf("referenced type '%s' must be capitalized", typeVal)
+					} else {
+						return fmt.Errorf("unknown atomic type '%s'", typeVal)
+					}
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// isStandardType checks if the given type is a EIP712 conformant type
+func isStandardTypeStr(encType string) bool {
+	// Atomic types
+	exp, _ := regexp.Compile(`^(address|bool|bytes|string)$`)
```

If we need to use regexps, it's better to use `MustCompile` if you're going to ignore the error anyway (and you are compiling a compile-time constant)